### PR TITLE
Push executor metrics

### PR DESCRIPTION
**Pull request by eseliger**
Local dashboard:

Screenshot 2022-06-12 at 18 29 51@2x

Follow-up tasks:
- [x] Implement push-through metrics for node_exporter on executors host
- [x] Implement push-through metrics for node_exporter on registry host
- [x] Expose from worker, not frontend
- [x] Remove metrics setup from docs
- [ ] Clean up terraform modules
  - [x] Remove credentials
  - [x] Remove open port
  - [x] Remove HTTP firewall rule
- [ ] Clean up k8s prometheus config
- [ ] Clean up Cloud prometheus config
- [ ] Clean up k8s secrets
- [ ] Clean up Cloud secrets
- [ ] Port this change to MI V1.0 and MI V1.1

### Partner PRs

- https://github.com/sourcegraph/deploy-sourcegraph-docker/pull/836
- https://github.com/sourcegraph/deploy-sourcegraph-helm/pull/147
- https://github.com/sourcegraph/deploy-sourcegraph/pull/4149

## Test plan

Validated locally that it still works.

### CODE REVIEW

File: docker-images/prometheus/config/prometheus.yml [hunk 1]
```diff
@@ -28,6 +28,12 @@ scrape_configs:
     metrics_path: /alertmanager/metrics
     static_configs:
       - targets: ['127.0.0.1:9093']
+  - job_name: sourcegraph-executors
+    honor_labels: true
+    metrics_path: /metrics-executors
+    static_configs:
+      - targets:
+        - host.docker.internal:6063
```
Comment (eseliger) [inline docker-images/prometheus/config/prometheus.yml:36]: TODO: This needs to be the frontend pods.

File: internal/debugserver/debug.go [hunk 2]
```diff
@@ -14,11 +14,15 @@ import (
...
 	"golang.org/x/net/trace"
 
+	dto "github.com/prometheus/client_model/go"
+
 	"github.com/sourcegraph/sourcegraph/internal/env"
 	"github.com/sourcegraph/sourcegraph/internal/goroutine"
 	"github.com/sourcegraph/sourcegraph/internal/httpserver"
 )
 
+var CurrentMetrics map[string][]*dto.MetricFamily
```
Comment (efritz) [inline internal/debugserver/debug.go:24]: Are the changes in this file still relevant?

**Comment by Strum355** [comment]
- https://github.com/sourcegraph/terraform-aws-executors/pull/37
- https://github.com/sourcegraph/terraform-google-executors/pull/50

File: dev/gitserver/sg.config.yaml [hunk 3]
```diff
@@ -18,8 +18,7 @@ env:
       { "Name": "precise-code-intel-worker", "Host": "127.0.0.1:6088" },
       { "Name": "worker", "Host": "127.0.0.1:6089" },
       { "Name": "enterprise-worker", "Host": "127.0.0.1:6089" },
-      { "Name": "executor-codeintel", "Host": "127.0.0.1:6092" },
-      { "Name": "executor-batches", "Host": "127.0.0.1:6093" },
```
Comment (efritz) [inline dev/gitserver/sg.config.yaml:22]: Do we know what this file is for?

File: enterprise/cmd/executor/internal/apiclient/client.go [hunk 4]
```diff
@@ -264,3 +286,23 @@ func intsToString(ints []int) string {
 
 	return strings.Join(segments, ", ")
 }
+
+func gatherMetrics(ctx context.Context, gatherer prometheus.Gatherer) (string, error) {
+	mfs, err := gatherer.Gather()
+	if err != nil {
+		return "", err
+	}
```
Comment (efritz) [inline enterprise/cmd/executor/internal/apiclient/client.go:294]: Note: no context

File: monitoring/definitions/shared/codeintel.go [hunk 5]
```diff
@@ -338,10 +338,19 @@ func (codeIntelligence) NewExecutorExecutionRunLockContentionGroup(containerName
 		Hidden: true,
 		Rows: []monitoring.Row{
 			{
-				constructor("executor_run_lock_wait", "wait").WithNoAlerts(`
```
Comment (efritz) [inline monitoring/definitions/shared/codeintel.go:341]: Is constructor now unused?

File: dev/gitserver/sg.config.yaml (hunk 3 above)
Comment (eseliger) [reply dev/gitserver/sg.config.yaml:22]: no

File: enterprise/cmd/worker/internal/executors/metricsserver_job.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+func (j *metricsServerJob) Description() string {
+	return "HTTP server exposing the metrics collected from executors to Prometheus"
+}
+
+func (j *metricsServerJob) Config() []env.Config {
+	return []env.Config{metricsServerConfigInst}
+}
+
+func (j *metricsServerJob) Routines(ctx context.Context, logger log.Logger) ([]goroutine.BackgroundRoutine, error) {
+	addr := fmt.Sprintf(":%d", metricsServerConfigInst.MetricsServerPort)
```
Comment (efritz) [inline enterprise/cmd/worker/internal/executors/metricsserver_job.go:34]: Should we explicitly use 0.0.0.0?

**Review by efritz — APPROVED** [review]

File: dev/src-prof-services.json [hunk 7]
```diff
@@ -5,9 +5,8 @@
   { "Name": "symbols", "Host": "127.0.0.1:6071" },
   { "Name": "repo-updater", "Host": "127.0.0.1:6074" },
   { "Name": "worker", "Host": "127.0.0.1:6089" },
+  { "Name": "worker-executors", "Host": "127.0.0.1:6969" },
   { "Name": "precise-code-intel-worker", "Host": "127.0.0.1:6088" },
-  { "Name": "executor-codeintel", "Host": "127.0.0.1:6092" },
-  { "Name": "executor-batches", "Host": "127.0.0.1:6093" },
```
Comment (Strum355) [inline dev/src-prof-services.json:10]: Do we need to modify this? Id assume we still have pprof technically on the executors

Comment (eseliger) [reply dev/src-prof-services.json:10]: hmmm yeah, makes sense. I have stepped out for a bit, could you restore it and see if we can merge then, if you got time? :)

**Pull request merged**